### PR TITLE
http: adding CONNECT example configs and improving docs

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: n/a
Testing: n/a
Docs Changes: connect docs improved
Release Notes: n/a
Part of #1451 and #1630

### CODE REVIEW

File: configs/encapsulate_in_connect.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+admin:
```
Comment (mattklein123) [inline configs/encapsulate_in_connect.yaml:1]: drive by: please make sure these new tests are added to the BUILD file so they get tested as part of config_test. Thank you!

File: configs/encapsulate_in_connect.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    address:
+      socket_address:
+        protocol: TCP
+        address: 127.0.0.1
+        port_value: 10000
+    filter_chains:
+    - filters:
+      - name: tcp
+        typed_config:
+          "@type": type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy
```
Comment (moderation) [inline configs/encapsulate_in_connect.yaml:20]: Use v3? `"@type": "type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy",`

File: configs/terminate_connect.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      - lb_endpoints:
+        - endpoint:
+            address:
+              socket_address:
+                address: www.google.com
+                port_value: 443
+    transport_socket:
+      name: envoy.transport_sockets.tls
+      typed_config:
+        "@type": type.googleapis.com/envoy.api.v2.auth.UpstreamTlsContext
```
Comment (moderation) [inline configs/terminate_connect.yaml:63]: Use v3 config? `"@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext"`

File: docs/root/intro/arch_overview/http/upgrades.rst [hunk 4]
```diff
@@ -89,3 +89,17 @@ upgrade requests or responses with bodies.
 .. This mode of CONNECT support can create major security holes if configured correctly, as the upstream
 .. will be forwarded *unsanitized* headers if they are in the body payload. Please use with caution
 
+.. Tunneling TCP over HTTP/2
+.. ^^^^^^^^^^^^^^^^^^^^^^^^^
+.. Envoy also has support for transforming raw TCP into HTTP/2 CONNECT requests. This can be used to
+.. proxy multiplexed TCP over pre-warmed secure connections and amortize the cost of any TLS handshake.
+.. An example set up proxying SMTP would look something like this
+..
+.. [SMTP Upstream] --- raw STMP --- [L2 Envoy]  --- SMTP tunneled over HTTP/2  --- [L1 Envoy]  --- SMTP  --- [Client]
```
Comment (snowp) [inline docs/root/intro/arch_overview/http/upgrades.rst:98]: typo: `raw STMP` - > `raw SMTP`

File: test/config_test/config_test.cc [hunk 5]
```diff
@@ -168,6 +168,8 @@ uint32_t run(const std::string& directory) {
     ENVOY_LOG_MISC(info, "testing {}.\n", filename);
     OptionsImpl options(
         Envoy::Server::createTestOptionsImpl(filename, "", Network::Address::IpVersion::v6));
+    // Try to avoid conflicting with other tests.
```
Comment (snowp) [inline test/config_test/config_test.cc:171]: should we just disable hot restart instead? or do some of the tests rely on that?

File: configs/encapsulate_in_connect.v3.yaml [hunk 6]
```diff
@@ -0,0 +1,38 @@
+admin:
```
Comment (snowp) [inline configs/encapsulate_in_connect.v3.yaml:1]: do we have docs explaining how this feature works as well?

Comment (alyssawilk) [reply configs/encapsulate_in_connect.v3.yaml:1]: Just what's in this PR.  I think folks will mainly only use this for TCP tunneling - while there is a desire to have Envoy generate CONNECT requests AFIK it's only for HTTP/1.1, where for HTTP/2 the only planned use is tunneling TCP (by Google and folks who asked for https://github.com/envoyproxy/envoy/issues/1630)

**Comment by alyssawilk** [comment]
Any other requests?

File: test/config_test/config_test.cc [hunk 7]
```diff
@@ -168,6 +168,8 @@ uint32_t run(const std::string& directory) {
     ENVOY_LOG_MISC(info, "testing {}.\n", filename);
     OptionsImpl options(
         Envoy::Server::createTestOptionsImpl(filename, "", Network::Address::IpVersion::v6));
+    // Avoid contention issues with other tests over the hot restart domain socket.
+    options.setHotRestartDisabled(true);
```
Comment (mattklein123) [inline test/config_test/config_test.cc:172]: Hmm @zuercher is this the same issue you are trying to fix?

Comment (mattklein123) [reply test/config_test/config_test.cc:172]: See https://github.com/envoyproxy/envoy/pull/11016/files. I'm guessing this change can be reverted once the other PR lands.

**Pull request merged**